connector/dynamicroutingconnector/README.md

```diff
@@ -0,0 +1,288 @@
+| `pipelines` | [][]pipeline.ID | Array of pipeline ID arrays, one for each threshold bucket. The number of pipeline arrays must be `len(thresholds) + 1` (to include the +inf bucket). | Yes |
+| `default_pipelines` | []pipeline.ID | Pipelines to use when the primary metadata key is missing from the client context. | No |
+| `evalaution_interval` | duration | How often to re-evaluate routing decisions based on new cardinality estimates. Default: 30s | No |
+
+### Configuration Rules
+
+- `thresholds` must be in ascending order and contain unique values
+- The number of `pipelines` must equal `len(thresholds) + 1` (one for each threshold interval plus one for values above the highest threshold)
+- At least one pipeline must be defined
+
+### Routing Logic
+
+The connector routes data based on the estimated cardinality for the primary key:
+
+- If `primary_metadata_key` is missing → routes to `default_pipelines`
+- If cardinality < `thresholds[0]` → routes to `pipelines[0]`
+- If `thresholds[i]` ≤ cardinality < `thresholds[i+1]` → routes to `pipelines[i+1]`
+- If cardinality ≥ `thresholds[len(thresholds)-1]` → routes to `pipelines[len(pipelines)-1]`
+
+## Use Cases
+
+### Dynamic Batching Based on Cardinality
+
+One of the most powerful use cases for the Dynamic Routing Connector is implementing dynamic batching strategies based on cardinality. By configuring `metadata_keys` to represent unique connections (e.g., source IP and user agent), you can route tenants with different connection volumes to different batching pipelines.
+
+**Scenario**: You're operating a multi-tenant observability platform where different tenants have vastly different cardinality patterns. Some tenants have a few unique combinations (e.g., few connections, few pods, few services), while others have many.
+
+**Problem**: Using a single batching configuration for all tenants leads to:
+- **Low-cardinality tenants**: Small batches that are inefficient and increase overhead
+- **High-cardinality tenants**: Large batches that may cause memory pressure and latency spikes
+
+**Solution**: Use the Dynamic Routing Connector to route tenants to different pipelines with optimized batching configurations based on their cardinality:
+
+```yaml
+receivers:
+  otlp:
+    protocols:
+      grpc:
+        endpoint: 0.0.0.0:4317
+
+connectors:
+  dynamicrouting:
+    primary_metadata_key: "x-tenant-id"
+    metadata_keys:
+      - "x-forwarded-for"
+      - "user-agent"
+    thresholds: [10, 50, 200]
+    pipelines:
+      # 0-10 unique connections: Small batches, frequent flush
+      - ["traces/small_batch"]
+      # 11-50 unique connections: Medium batches
+      - ["traces/medium_batch"]
+      # 51-200 unique connections: Large batches
+      - ["traces/large_batch"]
+      # 200+ unique connections: Very large batches, aggressive batching
+      - ["traces/xlarge_batch"]
```

Comment:
```suggestion
    dynamic_pipelines:
      # 0-10 unique connections: Small batches, frequent flush
      - max_count: 10
        pipelines: ["traces/small_batch"]
      # 11-50 unique connections: Medium batches
      - max_count: 50
        pipelines: ["traces/medium_batch"]
      # 51-200 unique connections: Large batches
      - max_count: 200
        pipelines: ["traces/large_batch"]
      # 200+ unique connections: Very large batches, aggressive batching
      - pipelines: ["traces/xlarge_batch"]
```

WDYT? We can have validation that there's exactly one entry without "max" defined, which will be used for the +Inf case.

Reply:
Like it! updated the code.